server/src/main/java/org/elasticsearch/cluster/metadata/SystemIndexMetadataUpgradeService.java

```diff
@@ -72,6 +74,35 @@ public void clusterChanged(ClusterChangedEvent event) {
         }
     }
 
+    // package-private for testing
+    boolean requiresUpdate(IndexMetadata indexMetadata) {
+        final boolean shouldBeSystem = shouldBeSystem(indexMetadata);
+
+        // should toggle system index status
+        if (shouldBeSystem != indexMetadata.isSystem()) {
+            return true;
+        }
+
+        if (shouldBeSystem) {
+            return shouldBeHidden(indexMetadata) || shouldHideAliases(indexMetadata);
+        }
+
+        return false;
+    }
+
+    private boolean shouldBeHidden(IndexMetadata indexMetadata) {
```

Comment:
Since these boolean logic methods are now nicely extracted, I think it would be really good if we added unit tests for the true false conditions. `shouldBeHidden` is perhaps trivial, but the others have certain level of control flow in them.

Reply:
Good idea! I added some tests that check the methods and whether they're interacting with requiresUpdate as we expect.